### DISCUSSION TITLE
(How) Can I duplicate a given project?

### DISCUSSION THREAD

**Discussion by astriffe** [zitadel/zitadel#1896] [Q&A]
Assume I have a project configured for the dev stage. 
How can I duplicate the project with all attached clients and only re-configure the changes required for additional stages?

Thanks, 
Alex

**Comment by livio-a** [comment]
Hi @astriffe 

At the moment you would have to do this manually by creating a new project with roles, applications and so on.

I see your point. Let's say you have a project with 3 applications (web, native api), some roles  and lots of authorizations. Although it's more or less a one time thing, manual copying might take some time and is error-prone.

Do you already have a process or some requirements in mind?

IMO just plain copying will not be enough. Some might just want to copy apps and roles, but no authorizations.
There might be settings (e.g. dev mode, redirect_uris on apps) which differ in production. You can of course change them afterwards, but this is error-prone as well and might lead to security risks.

**Reply by astriffe** [reply]
Thanks a lot @livio-a for your answer and your thougts. And please excuse my delayed reaction -- summer vacation ;-)

I could imagine a wizard (like you have it for application creation) would be helpful, where you can: 
* tick whether or not to keep development mode
* define the 'search-replace' for the url patterns and
* check or uncheck which roles you want to copy (and which ones you don't)

Alternatively, for me it would be evenly convenient to have an export and import functionality, meaning: I can export my dev prroject, adapt settings, create the new stage and import these settings. This would even enable a (poor man's approach) of a kind-of version controlled configuration... :)

What do you think?

**Reply by livio-a** [reply]
Hey @astriffe

I get the idea and agree on the "app creation" like wizard.

From a technical point I'm trying to decide whether we should implement this just as a UI process or have a specific API call as well.
The former would lack transaction handling, meaning if you encounter an error, only part of the copy process might have succeeded and you would have to delete it and retry again or do the rest manually again.
The later would of course solve this issue. But at the moment we don't have "use case" specific calls. But if we provide a possibility to "clone" a project, I'd prefer a specific API call.

@adlerhurst @fgerschwiler @fforootd any objections or additions?

**Reply by fforootd** [reply]
If we are going to make such a copy process I would prefer to implement it as an API function and not in the GUI only 😁

I personally find this feature interesting, so that someone can create a dev project and then clone it as base for a production project.

**Reply by adlerhurst** [reply]
I'm thinking of the following question which we should discuss for a use-cases API especially when coping objects/events:

Will we create a new event type or will we duplicate all events of the given object('s)? I would prefer a new event type.

Which dependencies/sub-objects should we copy?  
In the case of `project's` will we only copy `project's`, `application's`, `role's` or will we also copy `grant's` and `member's`?
